cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -0,0 +1,332 @@
+  uint32_t const num_poly_idxs,         // number of polyline indices
+  uint32_t const *poly_idxs,
+
+  cudf::column_device_view quad_lengths,  // numbers of points in each quadrant
+  cudf::column_device_view quad_offsets,  // offset of first point in each quadrant
+  cudf::column_device_view point_indices,
+  cudf::column_device_view point_x,
+  cudf::column_device_view point_y,
+
+  cudf::column_device_view poly_offsets,  // positions of the first vertex in a polyline
+  cudf::column_device_view poly_points_x,
+  cudf::column_device_view poly_points_y,
+
+  cudf::mutable_column_device_view out_point_index,
+  cudf::mutable_column_device_view out_poly_index,
+  cudf::mutable_column_device_view out_distance)
+{
+  // each block processes a quadrant
+  auto quad_pos = blockIdx.x + gridDim.x * blockIdx.y;
+  auto quad_idx = quad_idxs[quad_pos];
+
+  auto poly_begin = poly_idx_offsets[quad_pos];
+  auto poly_end =
+    quad_pos < num_poly_idx_offsets - 1 ? poly_idx_offsets[quad_pos + 1] : num_poly_idxs;
+
+  auto num_points = quad_lengths.element<uint32_t>(quad_idx);
+
+  for (auto tid = threadIdx.x; tid < num_points; tid += blockDim.x) {
+    // each thread loads its point
+    auto point_pos       = quad_offsets.element<uint32_t>(quad_idx) + tid;
+    auto point_id        = point_indices.element<uint32_t>(point_pos);
+    auto px              = point_x.element<T>(point_id);
+    auto py              = point_y.element<T>(point_id);
+    T distance           = 1e20;
+    auto nearest_poly_id = static_cast<uint32_t>(-1);
+
+    for (uint32_t poly_id = poly_begin; poly_id < poly_end; poly_id++)  // for each polyline
+    {
+      auto poly_idx   = poly_idxs[poly_id];
+      auto ring_begin = poly_offsets.element<uint32_t>(poly_idx);
+      auto ring_end   = poly_idx < poly_offsets.size() - 1
+                        ? poly_offsets.element<uint32_t>(poly_idx + 1)
+                        : poly_points_x.size();
+      auto ring_len = ring_end - ring_begin;
+
+      for (auto point_idx = 0; point_idx < ring_len; point_idx++)  // for each line
+      {
+        T x0  = poly_points_x.element<T>(ring_begin + ((point_idx + 0) % ring_len));
+        T y0  = poly_points_y.element<T>(ring_begin + ((point_idx + 0) % ring_len));
+        T x1  = poly_points_x.element<T>(ring_begin + ((point_idx + 1) % ring_len));
+        T y1  = poly_points_y.element<T>(ring_begin + ((point_idx + 1) % ring_len));
+        T d1x = x1 - x0, d1y = y1 - y0;
+        T d2x = px - x0, d2y = py - y0;
+        T d1 = sqrt(d1x * d1x + d1y * d1y);
+        T r  = (d1x * d2x + d1y * d2y) / d1;
+        T d  = 1e20;
+        if (r <= 0 || r >= d1) {
+          T d3x = px - x1, d3y = py - y1;
+          T d2 = sqrt(d2x * d2x + d2y * d2y);
+          T d3 = sqrt(d3x * d3x + d3y * d3y);
+          d    = min(min(d, d2), d3);
+        } else {
+          d = sqrt((d2x * d2x + d2y * d2y) - (r * r));
+        }
```

Comment:
For a function that only needs to find the nearest distance, this is doing a lot of expensive sqrts. 

If you only want to find the minimum distance, you should only need to calculate *distance squared*. Then, once you find the minimum distance squared, take the sqrt of just that distance-squared to find the minimum distance.

I bet this will be a lot faster...